### ISSUE TITLE
race between pool and PV creation

### ISSUE THREAD

**Issue by bassam** [rook/rook#767]
The integration tests create the following:

```
apiVersion: rook.io/v1alpha1
kind: Pool
metadata:
  name: replicapool
  namespace: rook
spec:
  replication:
    size: 1
  # For an erasure-coded pool, comment out the replication count above and uncomment the following settings.
  # Make sure you have enough OSDs to support the replica count or erasure code chunks.
  #erasureCode:
  #  codingChunks: 2
  #  dataChunks: 2
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: rook-block
provisioner: rook.io/block
parameters:
    pool: replicapool
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: block-pv-claim
  annotations:
    volume.beta.kubernetes.io/storage-class: rook-block
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1M
```

but the PersistentVolumeClaim seems to be created before the Pool is created leading in a failure:

```
> kubectl logs rook-api-2485995279-d1pvp --namespace rook | grep replica
2017-06-28 08:00:54.824596 I | exec: Running command: rbd create replicapool/k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e94b5d5b-5bd7-11e7-855c-0a580ac00302 --size 1 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring
2017-06-28 08:00:54.932104 E | api: failed to create image {Name:k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e94b5d5b-5bd7-11e7-855c-0a580ac00302 PoolName:replicapool Size:1000000 Device: MountPoint:}: failed to create image k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e94b5d5b-5bd7-11e7-855c-0a580ac00302 in pool replicapool of size 1048576: Failed to complete : exit status 2. output: rbd: error opening pool replicapool: (2) No such file or directory
2017-06-28 08:00:55.137547 I | exec: Running command: ceph osd pool create replicapool 0 replicated --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/480999081
2017-06-28 08:00:55.281874 I | exec: Running command: rbd create replicapool/k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e9915691-5bd7-11e7-855c-0a580ac00302 --size 1 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring
2017-06-28 08:00:55.317201 E | api: failed to create image {Name:k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e9915691-5bd7-11e7-855c-0a580ac00302 PoolName:replicapool Size:1000000 Device: MountPoint:}: failed to create image k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-e9915691-5bd7-11e7-855c-0a580ac00302 in pool replicapool of size 1048576: Failed to complete : exit status 2. output: rbd: error opening pool replicapool: (2) No such file or directory
2017-06-28 08:00:55.731585 I | exec: pool 'replicapool' created
2017-06-28 08:00:55.732127 I | exec: Running command: ceph osd pool set replicapool size 1 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/939657716
2017-06-28 08:00:56.806107 I | cephclient: creating pool replicapool succeeded, buf: 
2017-06-28 08:01:10.281525 I | exec: Running command: rbd create replicapool/k8s-dynamic-pvc-e945d0af-5bd7-11e7-8ac4-02420ac00002-f28248d7-5bd7-11e7-855c-0a580ac00302 --size 1 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring
2017-06-28 08:01:11.907592 I | exec: Running command: rbd ls -l replicapool --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json
```

Not sure how we should handle this kind of race, but it would not be ideal if we had to ask the user to wait for pool creation and if this kind of dependency is handled by the operator.

@travisn @jbw976

**Comment by kokhang** [comment]
This due to the fact that the creation of the pool and creation of the image is in the same manifest.
When Pool CRDs are created, it is watched by the Rook and a request is sent to the Rook cluster to create it.

If we want to solve this, we can add status to the Pool CRD and set it to `ready` state when it is created. But i see this as a low priority since the admin is the one setting up the clusters and users are the one consuming it. I feel we just need to put the setting up of the cluster (creation of clusters, storageclass, pools) in a different manifest than the consumption of the storage (pvc, mysql pod).

**Comment by jbw976** [comment]
it's worth noting that all our documentation and examples are structured that way too, where the storageclass (and pool) are created before any PVCs (and in a separate manifest)

**Comment by DanKerns** [comment]
Let me just flag my concern around this.

I have deep deep scars from nearly every time I've made a timing assumption in my code. Just because we manually enter two manifests doesn't mean someone else won't this in a script or a single manifest and dramatically change the timing.

Ignoring a race condition, in my experience, just hides it until the worst possible time.

**Comment by kokhang** [comment]
I believe this is handled by Kubernetes's controller and state-watching. If a pvc gets created before a pool object, yes it will fail but the next time around, it would succeed.

**Comment by DanKerns** [comment]
If that was try, then why did it fail?  Let's get serious about fixing this.

**Comment by jbw976** [comment]
I believe the last 2 statements in the log above show the image (PV) being created successfully.  If we had more of the log here I think we'd see the rest of the flow did succeed OK after the eventual consistency is reached.

I feel confident that the machinery is in place to handle this scenario and that it's not a true race condition that will never be recovered from.  I believe this is more of an eventual consistency scenario where there is a missing dependency (the pool), but once that dependency gets fulfilled then the rest of the flow will succeed automatically.

**Comment by jbw976** [comment]
As discussed, the machinery should be in place to get the pod to a good state (PVC bound, mounted and pod running).  The integration tests should only wait on that final outcome, they shouldn't have to wait for the intermediate steps to get there, such as the PVC becoming bound.  @dangula will update the tests to give us confidence that we have the machinery in place to handle the timing and retries of the full end to end volume mounting.
